core/node/api_server/src/tx_sender/gas_estimation.rs

```diff
@@ -0,0 +1,511 @@
+    ExecuteTransactionCommon, PackedEthSignature, ProtocolVersionId, Transaction,
+};
+
+use super::{result::ApiCallResult, SubmitTxError, TxSender};
+use crate::execution_sandbox::{BlockArgs, SandboxAction, VmPermit, SANDBOX_METRICS};
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum BinarySearchKind {
+    /// Full binary search.
+    Full,
+    /// Binary search with an optimized initial pivot.
+    #[allow(dead_code)] // FIXME
+    Optimized,
+}
+
+impl TxSender {
+    #[tracing::instrument(level = "debug", skip_all, fields(
+        initiator = ?tx.initiator_account(),
+        nonce = ?tx.nonce(),
+    ))]
+    pub async fn get_txs_fee_in_wei(
+        &self,
+        tx: Transaction,
+        estimated_fee_scale_factor: f64,
+        acceptable_overestimation: u64,
+        state_override: Option<StateOverride>,
+        kind: BinarySearchKind,
+    ) -> Result<Fee, SubmitTxError> {
+        let estimation_started_at = Instant::now();
+        let mut estimator = GasEstimator::new(self, tx, state_override).await?;
+        estimator.adjust_transaction_fee();
+
+        let initial_estimate = estimator.initialize().await?;
+        tracing::trace!(
+            "preparation took {:?}, starting binary search",
+            estimation_started_at.elapsed()
+        );
+
+        let optimized_lower_bound = initial_estimate.lower_gas_bound_without_overhead();
+        // Perform an initial search iteration with the pivot slightly greater than `gas_used` to account for 63/64 rule for far calls etc.
+        // If the transaction succeeds, it will discard most of the search space at once.
+        let optimistic_gas_limit = initial_estimate.optimistic_gas_limit_without_overhead();
+
+        let (bounds, initial_pivot) = match kind {
+            BinarySearchKind::Full => {
+                let lower_bound = initial_estimate.gas_charged_for_pubdata;
+                let upper_bound = MAX_L2_TX_GAS_LIMIT;
+                (lower_bound..=upper_bound, None)
+            }
+            BinarySearchKind::Optimized => {
+                let lower_bound = optimized_lower_bound.unwrap_or(0);
+                let upper_bound = MAX_L2_TX_GAS_LIMIT;
+                let initial_pivot = optimistic_gas_limit.filter(|&gas| {
+                    // If `optimistic_gas_limit` is greater than the ordinary binary search pivot, there's no sense using it.
+                    gas < (lower_bound + upper_bound) / 2
+                });
+                (lower_bound..=upper_bound, initial_pivot)
+            }
+        };
+
+        let unscaled_gas_limit =
+            Self::binary_search(&estimator, bounds, initial_pivot, acceptable_overestimation)
+                .await?;
+        // Metrics are intentionally reported regardless of the binary search mode, so that the collected stats can be used to adjust
+        // optimized binary search params (e.g., the initial pivot multiplier).
```

Comment:
Does it make sense to log the lower bound / initial gas limit, or would it produce too much spam?

Reply:
I would log it TBH - it's less than ~30 per second currently, will not make an impact on the total API logs